package dev.by1337.virtualentity.core.virtual.npc;

import dev.by1337.core.ServerVersion;
import dev.by1337.virtualentity.api.annotations.SinceMinecraftVersion;
import dev.by1337.virtualentity.api.entity.VirtualEntityType;
import dev.by1337.virtualentity.api.entity.npc.VillagerData;
import dev.by1337.virtualentity.api.entity.npc.VillagerProfession;
import dev.by1337.virtualentity.api.entity.npc.VillagerType;
import dev.by1337.virtualentity.core.mappings.Mappings;
import dev.by1337.virtualentity.core.syncher.EntityDataAccessor;

public class VirtualVillagerImpl extends VirtualAbstractVillagerImpl implements dev.by1337.virtualentity.api.virtual.npc.VirtualVillager {
    private static final EntityDataAccessor<VillagerData> DATA_VILLAGER_DATA;
    @SinceMinecraftVersion("26.1")
    private static final EntityDataAccessor<Boolean> DATA_VILLAGER_DATA_FINALIZED;

    static {
        DATA_VILLAGER_DATA = Mappings.findAccessor("Villager", "DATA_VILLAGER_DATA");
        if (ServerVersion.is26_1orNewer()) {
            DATA_VILLAGER_DATA_FINALIZED = Mappings.findAccessor("Villager", "DATA_VILLAGER_DATA_FINALIZED");
        } else {
            DATA_VILLAGER_DATA_FINALIZED = null;
        }
    }

    public VirtualVillagerImpl() {
        super(VirtualEntityType.VILLAGER);
    }

    protected void defineSynchedData() {
        super.defineSynchedData();
        this.entityData.define(DATA_VILLAGER_DATA, new VillagerData(VillagerType.PLAINS, VillagerProfession.NONE, 1));
        if (ServerVersion.is26_1orNewer()) {
            entityData.define(DATA_VILLAGER_DATA_FINALIZED, false);
        }
    }

    @SinceMinecraftVersion("26.1")
    public boolean isDataFinalized() {
        return entityData.get(DATA_VILLAGER_DATA_FINALIZED);
    }

    @SinceMinecraftVersion("26.1")
    public void setDataFinalized(boolean value) {
        entityData.set(DATA_VILLAGER_DATA_FINALIZED, value);
    }

    /**
     * Получает данные о деревенском жителе.
     * Включает тип деревни, профессию и уровень.
     *
     * @return объект {@link VillagerData}, который содержит информацию о деревенском жителе.
     */
    @Override
    public VillagerData getVillagerData() {
        return this.entityData.get(DATA_VILLAGER_DATA);
    }

    /**
     * Устанавливает данные о деревенском жителе.
     *
     * @param villagerData объект {@link VillagerData}, который содержит информацию о деревенском жителе.
     */
    @Override
    public void setVillagerData(VillagerData villagerData) {
        this.entityData.set(DATA_VILLAGER_DATA, villagerData);
    }
}
